py.path.local warning rasied on pytest 7rc1

```
============================================================= test session starts =============================================================
platform linux -- Python 3.10.2, pytest-7.0.0rc1, pluggy-1.0.0
Matplotlib: 3.5.1
Freetype: 2.6.1
rootdir: /home/stuart/Git/DKIST/dkist, configfile: setup.cfg
plugins: remotedata-0.3.2, openfiles-0.5.0, mock-3.6.1, filter-subpackage-0.1.1, doctestplus-0.11.2, cov-3.0.0, astropy-header-0.1.2, arraydiff-0.3, hypothesis-6.31.6, mpl-0.13, httpserver-1.0.3, anyio-3.4.0, forked-1.4.0, xdist-2.5.0, asdf-2.8.4.dev42+gef95881
collected 0 items / 1 error                                                                                                                   

=================================================================== ERRORS ====================================================================
________________________________________________________ ERROR collecting test session ________________________________________________________
../../../.virtualenvs/dkist/lib/python3.10/site-packages/pluggy/_hooks.py:265: in __call__
    return self._hookexec(self.name, self.get_hookimpls(), kwargs, firstresult)
../../../.virtualenvs/dkist/lib/python3.10/site-packages/pluggy/_manager.py:80: in _hookexec
    return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
../../asdf/pytest_asdf/plugin.py:318: in pytest_collect_file
    return AsdfSchemaFile.from_parent(
../../asdf/pytest_asdf/plugin.py:62: in from_parent
    result = super().from_parent(parent, fspath=fspath, **kwargs)
../../../.virtualenvs/dkist/lib/python3.10/site-packages/_pytest/nodes.py:630: in from_parent
    return super().from_parent(parent=parent, fspath=fspath, path=path, **kw)
../../../.virtualenvs/dkist/lib/python3.10/site-packages/_pytest/nodes.py:261: in from_parent
    return cls._create(parent=parent, **kw)
../../../.virtualenvs/dkist/lib/python3.10/site-packages/_pytest/nodes.py:140: in _create
    return super().__call__(*k, **kw)
../../../.virtualenvs/dkist/lib/python3.10/site-packages/_pytest/nodes.py:585: in __init__
    path = _imply_path(type(self), path, fspath=fspath)
../../../.virtualenvs/dkist/lib/python3.10/site-packages/_pytest/nodes.py:110: in _imply_path
    warnings.warn(
E   pytest.PytestRemovedIn8Warning: The (fspath: py.path.local) argument to AsdfSchemaFile is deprecated. Please use the (path: pathlib.Path) argument instead.
E   See https://docs.pytest.org/en/latest/deprecations.html#fspath-argument-for-node-constructors-replaced-with-pathlib-path
=========================================================== short test summary info ===========================================================
ERROR  - pytest.PytestRemovedIn8Warning: The (fspath: py.path.local) argument to AsdfSchemaFile is deprecated. Please use the (path: pathlib...
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! Interrupted: 1 error during collection !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
============================================================== 1 error in 0.13s ===============================================================
```

--- a/CHANGES.rst
+++ b/CHANGES.rst
@@ -1,3 +1,8 @@
+2.9.2 (unreleased)
+------------------
+
+- Fix deprecation warnings stemming from the release of pytest 7.0.0. [#1075]
+
 2.9.1 (2022-02-03)
 ------------------
 

--- a/asdf/tests/helpers.py
+++ b/asdf/tests/helpers.py
@@ -370,14 +370,18 @@ def assert_no_warnings(warning_class=None):
         emitted.
     """
     import pytest
-    with pytest.warns(None) as recorded_warnings:
-        yield
 
-    if warning_class is not None:
+    if warning_class is None:
+        with warnings.catch_warnings():
+            warnings.simplefilter("error")
+
+            yield
+    else:
+        with pytest.warns(Warning) as recorded_warnings:
+            yield
+
         assert not any(isinstance(w.message, warning_class) for w in recorded_warnings), \
             display_warnings(recorded_warnings)
-    else:
-        assert len(recorded_warnings) == 0, display_warnings(recorded_warnings)
 
 
 def assert_extension_correctness(extension):

--- a/asdf/tests/test_types.py
+++ b/asdf/tests/test_types.py
@@ -385,7 +385,7 @@ undefined_data:
         - !core/complex-1.0.0 3.14j
 """
     buff = helpers.yaml_to_asdf(yaml)
-    with pytest.warns(None) as warning:
+    with pytest.warns(Warning) as warning:
         afile = asdf.open(buff)
         missing = afile.tree['undefined_data']
 

--- a/pytest_asdf/plugin.py
+++ b/pytest_asdf/plugin.py
@@ -60,12 +60,21 @@ class AsdfSchemaFile(pytest.File):
     def from_parent(cls, parent, *, fspath, skip_examples=False, validate_default=True,
         ignore_unrecognized_tag=False, ignore_version_mismatch=False, skip_tests=[], xfail_tests=[], **kwargs):
 
-        path = pathlib.Path(fspath)
+        from asdf.util import minversion
+
+        if minversion("pytest", "7.0.0"):
+            path = pathlib.Path(fspath)
+
+            kwargs["path"] = path
+        else:
+            path = fspath
+
+            kwargs["fspath"] = path
 
         if hasattr(super(), "from_parent"):
-            result = super().from_parent(parent, path=path, **kwargs)
+            result = super().from_parent(parent, **kwargs)
         else:
-            result = AsdfSchemaFile(path, parent, **kwargs)
+            result = AsdfSchemaFile(path, parent)
 
         result.skip_examples = skip_examples
         result.validate_default = validate_default

--- a/pytest_asdf/plugin.py
+++ b/pytest_asdf/plugin.py
@@ -1,5 +1,6 @@
 import io
 import os
+import warnings
 from importlib.util import find_spec
 from pkg_resources import parse_version
 import pathlib
@@ -64,7 +65,7 @@ class AsdfSchemaFile(pytest.File):
         if hasattr(super(), "from_parent"):
             result = super().from_parent(parent, path=path, **kwargs)
         else:
-            result = AsdfSchemaFile(fspath, parent, **kwargs)
+            result = AsdfSchemaFile(path, parent, **kwargs)
 
         result.skip_examples = skip_examples
         result.validate_default = validate_default
@@ -241,10 +242,11 @@ class AsdfSchemaExampleItem(pytest.Item):
         b._array_storage = "streamed"
 
         try:
-            with pytest.warns(None) as w:
-                ff._open_impl(ff, buff, mode='rw')
             # Do not tolerate any warnings that occur during schema validation
-            assert len(w) == 0, helpers.display_warnings(w)
+            with warnings.catch_warnings():
+                warnings.simplefilter("error")
+
+                ff._open_impl(ff, buff, mode='rw')
         except Exception:
             print("From file:", self.filename)
             raise

--- a/pytest_asdf/plugin.py
+++ b/pytest_asdf/plugin.py
@@ -58,8 +58,11 @@ class AsdfSchemaFile(pytest.File):
     @classmethod
     def from_parent(cls, parent, *, fspath, skip_examples=False, validate_default=True,
         ignore_unrecognized_tag=False, ignore_version_mismatch=False, skip_tests=[], xfail_tests=[], **kwargs):
+
+        path = pathlib.Path(fspath)
+
         if hasattr(super(), "from_parent"):
-            result = super().from_parent(parent, fspath=fspath, **kwargs)
+            result = super().from_parent(parent, path=path, **kwargs)
         else:
             result = AsdfSchemaFile(fspath, parent, **kwargs)
 

--- a/pytest_asdf/plugin.py
+++ b/pytest_asdf/plugin.py
@@ -60,15 +60,13 @@ class AsdfSchemaFile(pytest.File):
     def from_parent(cls, parent, *, fspath, skip_examples=False, validate_default=True,
         ignore_unrecognized_tag=False, ignore_version_mismatch=False, skip_tests=[], xfail_tests=[], **kwargs):
 
+        # Fix for depreciation of fspath in pytest 7+
         from asdf.util import minversion
-
         if minversion("pytest", "7.0.0"):
             path = pathlib.Path(fspath)
-
             kwargs["path"] = path
         else:
             path = fspath
-
             kwargs["fspath"] = path
 
         if hasattr(super(), "from_parent"):

